packages/webcomponentsjs/ts_src/platform/event-listener-options.ts

```diff
@@ -0,0 +1,123 @@
+ * Google as part of the polymer project is also subject to an additional IP
+ * rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+export {};
+
+// Older browsers like IE do not support an object as the options parameter
+// to add/removeEventListener.
+// https://connect.microsoft.com/IE/feedback/details/790389/event-defaultprevented-returns-false-after-preventdefault-was-called
+const supportsEventOptions = (() => {
+  let supported = false;
+  const eventOptions = {
+    get capture() {
+      supported = true;
+      return false;
+    },
+  };
+  const listener = () => {};
+  // NOTE: These will be unpatched at this point.
+  window.addEventListener('test', listener, eventOptions);
+  window.removeEventListener('test', listener, eventOptions);
+  return supported;
+})();
+
+const nativeEventTarget = window.EventTarget ?? window.Node;
+
+if (
+  !supportsEventOptions &&
+  'addEventListener' in nativeEventTarget.prototype
+) {
+  const parseEventOptions = (
+    optionsOrCapture?: boolean | AddEventListenerOptions
+  ) => {
+    let capture, once;
+    if (
+      optionsOrCapture &&
+      (typeof optionsOrCapture === 'object' ||
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        (optionsOrCapture as any) instanceof Object)
+    ) {
+      capture = Boolean((optionsOrCapture as AddEventListenerOptions).capture);
+      once = Boolean((optionsOrCapture as AddEventListenerOptions).once);
+    } else {
+      capture = Boolean(optionsOrCapture);
+      once = false;
+    }
+    return {
+      capture,
+      once,
+    };
+  };
+
+  const origAddEventListener = nativeEventTarget.prototype.addEventListener;
+  const origRemoveEventListener =
+    nativeEventTarget.prototype.removeEventListener;
+
+  const captureListenerMap = new WeakMap();
+  const listenerMap = new WeakMap();
+
+  const getListenerMap = (
+    target: EventTarget,
+    type: string,
+    capture: boolean
+  ) => {
+    const elMap = capture ? captureListenerMap : listenerMap;
+    let typeMap = elMap.get(target);
+    if (typeMap === undefined) {
+      elMap.set(target, (typeMap = new Map()));
+    }
+    let listeners = typeMap.get(type);
+    if (listeners === undefined) {
+      typeMap.set(type, (listeners = new WeakMap()));
+    }
+    return listeners;
+  };
+
+  nativeEventTarget.prototype.addEventListener = function (
+    type: string,
+    listener: EventListenerOrEventListenerObject | null,
+    options?: boolean | AddEventListenerOptions | undefined
+  ) {
+    if (listener == null) {
+      return;
+    }
+    const {capture, once} = parseEventOptions(options);
+    const map = getListenerMap(this, type, capture);
+    if (!map.has(listener)) {
+      const cachedListener = once
+        ? (e: Event) => {
+            map.delete(listener);
+            origRemoveEventListener.call(this, type, cachedListener, capture);
+            ((listener as EventListenerObject).handleEvent ?? listener)(e);
```

Comment:
I think this needs to be called with `call` to set up the correct `this`: `this`, when `listener` is the function being called, and `e.currentTarget`, when calling `listener.handleEvent`: if I'm reading the spec correctly, the DOM normally calls `handleEvent` as part of ['inner invoke'](https://dom.spec.whatwg.org/#concept-event-listener-inner-invoke), which calls ['call a user object's operation'](https://heycam.github.io/webidl/#call-a-user-objects-operation) by passing the event's `currentTarget` as `thisArg`. 'Call a user object's operation' also first checks if the callback object it's given is itself callable (calling that, if so) before falling back to trying to read `opName` (`"handleEvent"`).